[Op][O1][O2a] Utility, arithmetic and comparison operators

MasterJH5574 opened the pull request:
As tracked by #332, this PR is the kickoff part of high-level operator introduction in Relax.

This PR is about the milestone O1 and O2a. Specifically, this PR
* introduces some of common utility functions that the registration and StructInfo inference of each operator will often use.
* introduces unary arithmetic operators
  * `cos`, `log`, `negative`, `sigmoid`, `sin`, `sqrt`, `tanh`.
* refactors and introduces binary arithmetic operators
  * `add`, `divide`, `floor_divide`, `multiply`, `subtract`.
* introduces binary comparative operators
  * `equal`, `greater`, `greater_equal`, `less`, `less_equal`, `not_equal`.

These operators are well tested from three perspective:
* P1. the op getter can get correct op by name
* P2. their StructInfo inference result are as expected under all kinds of cases
* P3. Relax TVMScript parser can parse the scripts with the op inside

For operators in O2a, most operators share almost the same StructInfo inference logic. Therefore, for tests in P2, in each category, not every op is tested in every case. For each case, it is good to have only part of op in this category tested. This is intended not to make overlarge testing file.

Hzfengsy commented on python/tvm/relax/op/binary.py:
```diff
@@ -0,0 +1,232 @@
+    return _ffi_api.floor_divide(lhs, rhs)  # type: ignore
+
+
+def multiply(lhs: Expr, rhs: Expr) -> Expr:
```
Given PyTorch uses `mul` as its default API, and an alias for `multiply`. (https://pytorch.org/docs/stable/generated/torch.multiply.html?highlight=multiply#torch.multiply) How about changing to keep the same as PyTorch? i.e., 

```python
def mul(lhs: Expr, rhs: Expr):
    ...

multiply = mul
```

Hzfengsy replied:
FYI, Numpy and Tensorflow do not have `mul` API, but only have `multiply`. So, I don't have a strong opinion on this.

tqchen replied:
Let us go with numpy for now. Since the convention is clear. We can introduce a mul alias here

MasterJH5574 replied:
Let’s go with `multiply` first.

Hzfengsy commented on python/tvm/relax/op/unary.py:
```diff
@@ -0,0 +1,156 @@
+    return _ffi_api.log(data)  # type: ignore
+
+
+def negative(lhs: Expr) -> Expr:
```
```suggestion
def negative(data: Expr) -> Expr:
```

MasterJH5574 replied:
Thanks the catch!

sunggg reviewed (commented):
Thank you for the great work! 
Overall, LGTM and have a few questions below.

MasterJH5574 commented on python/tvm/relax/op/binary.py:
```diff
@@ -0,0 +1,232 @@
+###################### Arithmetic operators ######################
+
+
+def add(lhs: Expr, rhs: Expr) -> Expr:
```
Thanks for bringing this up! DataAPI uses `x1` and `x2` for binary ops and `x` for unary ops. https://data-apis.org/array-api/latest/API_specification/generated/array_api.add.html

I will do a round of renaming to align the specification.

MasterJH5574 commented on src/relax/op/op_common.h:
```diff
@@ -29,34 +29,125 @@
+ * be prepended with a prefix "relax." as the identifier string in the operator registry.
+ * \param RequireFloatDtype A boolean indicating if the input is required to have float dtype.
+ */
+#define RELAX_REGISTER_UNARY_OP(OpRegName, RequireFloatDtype) \
```
Oh yes they will be used by NN operators, specifically the activation functions like ReLU, GeLU, etc., whose identifier in registry is like `"relax.op.nn.relu"` while the make-function name should not contain the dot.

Given that they are not introduced in this PR, let me remove them first.

MasterJH5574 commented on src/relax/op/op_common.h:
```diff
@@ -29,34 +29,125 @@
+ * be prepended with a prefix "relax.op." as the FFI identifier string for the make function,
+ * \param OpRegName The identifier of the operator in the registry
+ */
+#define RELAX_UNARY_OP_IMPL(OpName, OpRegName)         \
```
Thanks for the comments. Do you think `RELAX_REGISTER_UNARY_OP_INTERFACE` makes sense? Though it is a bit long, but since the unary ops are all one-liners most of the time, that is not a big deal.

sunggg replied:
Thanks! Yeah, it is not a big deal, but in case you want to make it shorter, I think we can simply call it `RELAX_REGISTER_UNARY_OP` and handle NN operators differently.

MasterJH5574 replied:
Another case will be operator `tril`/`triu` (https://data-apis.org/array-api/latest/API_specification/generated/array_api.tril.html), whose returned tensor always has the same StructInfo as the input one, which means it can also share the inference function `InferStructInfoUnary` by using `RELAX_REGISTER_UNARY_OP`. While on the other hand, these two operators have a nontrivial attribute, which prevents them from sharing a same pattern of interface. So I think having separate macro there in the future still helps.

sunggg replied:
Yeap, I wasn't opposing for having two separate macros :) Just tried to make common case shorter and special case longer haha but again, it is not a big deal now so I'm okay with current naming. Thank you for reflecting my comments!

MasterJH5574 commented:
@sunggg Thanks for the comments so far! Just addressed your comments. The new changes including:
- renaming `data` to `x` for unary ops, `lhs`, `rhs` to `x1`, `x2` for binary ops, and old names to `x1` `x2` `x3` for ewise-fma.
- combined the macros for unary ops. Will introduce them again next time when NN ops come in.
- renamed the macro to use the word “interface”.

sunggg commented:
Thanks @MasterJH5574 for introducing relax operators and thanks @Hzfengsy @tqchen for the review!
